crates/youki/src/commands/run.rs

```diff
@@ -19,9 +25,78 @@ pub fn run(args: Run, root_path: PathBuf, systemd_cgroup: bool) -> Result<()> {
         .validate_id()?
         .as_init(&args.bundle)
         .with_systemd(systemd_cgroup)
+        .with_detach(args.detach)
         .build()?;
 
     container
         .start()
-        .with_context(|| format!("failed to start container {}", args.container_id))
+        .with_context(|| format!("failed to start container {}", args.container_id))?;
+
+    if args.detach {
+        return Ok(0);
+    }
+
+    // Using `debug_assert` here rather than returning an error because this is
+    // a invariant. The design when the code path arrives to this point, is that
+    // the container state must have recorded the container init pid.
+    debug_assert!(
+        container.pid().is_some(),
+        "expects a container init pid in the container state"
+    );
+    handle_foreground(container.pid().unwrap())
+}
+
+fn handle_foreground(init_pid: Pid) -> Result<i32> {
+    // We mask all signals here and forward most of the signals to the container
+    // init process.
+    let signal_set = SigSet::all();
+    signal_set
+        .thread_set_mask()
+        .with_context(|| "failed to call pthread_sigmask")?;
+    loop {
+        match signal_set
+            .wait()
+            .with_context(|| "failed to call sigwait")?
+        {
+            signal::SIGCHLD => {
+                // Reap all child until either container init process exits or
+                // no more child to be reaped. Once the container init process
+                // exits we can then return.
+                loop {
+                    match waitpid(None, Some(WaitPidFlag::WNOHANG))? {
+                        WaitStatus::Exited(pid, status) => {
+                            if pid.eq(&init_pid) {
+                                return Ok(status);
+                            }
+
+                            // Else, some random child process exited, ignoring...
+                        }
+                        WaitStatus::Signaled(pid, signal, _) => {
+                            if pid.eq(&init_pid) {
+                                return Ok(signal as i32);
+                            }
+
+                            // Else, some random child process exited, ignoring...
+                        }
+                        WaitStatus::StillAlive => {
+                            // No more child to reap.
+                            break;
+                        }
+                        _ => {}
+                    }
+                }
+            }
+            signal::SIGURG => {
+                // In `runc`, SIGURG is used by go runtime and should not be forwarded to
+                // the container process. Here, we just ignore the signal.
+            }
+            signal::SIGWINCH => {
+                // TODO: resize the terminal
```

Comment:
May I ask you to create the issue and put the link?

Reply:
https://github.com/containers/youki/issues/1672